File: src/main/kotlin/pt/isel/ls/server/data/dataInterfaces/models/CardData.kt
package pt.isel.ls.server.data.dataInterfaces.models

import pt.isel.ls.server.data.transactionManager.transactions.TransactionCtx
import pt.isel.ls.server.utils.Card

interface CardData {

    fun createCard(
        idList: Int,
        idBoard: Int,
        name: String,
        description: String? = null,
        endDate: String? = null,
        ctx: TransactionCtx
    ): Int

    fun getCardsFromList(idList: Int, idBoard: Int, ctx: TransactionCtx): List<Card>

    fun getCard(idCard: Int, idBoard: Int, ctx: TransactionCtx): Card

    fun moveCard(
        idCard: Int,
        idBoard: Int,
        idList: Int,
        idListDst: Int,
        idx: Int,
        idxDst: Int,
        ctx: TransactionCtx
    )

    fun decreaseIdx(idList: Int, idx: Int, ctx: TransactionCtx)

    fun deleteCard(idCard: Int, idBoard: Int, ctx: TransactionCtx)

    fun deleteCards(idList: Int, ctx: TransactionCtx)

    fun archiveCards(idBoard: Int, idList: Int, ctx: TransactionCtx)

    fun getNextIdx(idList: Int, ctx: TransactionCtx): Int

    fun getCardCount(idBoard: Int, idList: Int, ctx: TransactionCtx): Int // can be deleted, it's not being used.

    fun getArchivedCards(idBoard: Int, ctx: TransactionCtx): List<Card>

    fun updateCard(card: Card, description: String?, endDate: String?, idList: Int?, archived: Boolean, ctx: TransactionCtx)
}

File: static-content/data/cardData.js
import {fetchReq} from "../utils.js";


async function createCard(idBoard, idList, name) {
   return await fetchReq(`board/${idBoard}/card`, "POST", {
           idList: idList,
           name: name,
           description: null,
           endDate: null
       })
}

async function getCard(idBoard, idCard) {
    return await fetchReq(`board/${idBoard}/card/${idCard}`,"GET")
}

async function updateCard(idBoard, idCard, archived, newDescription, newEndDate, idList) {
    await fetchReq(`board/${idBoard}/card/${idCard}/update`, "PUT", {
        archived: archived,
        description: newDescription,
        endDate: newEndDate,
        idList: idList
    })
}

async function deleteCard(idBoard, idCard) {
    await fetchReq(`board/${idBoard}/card/${idCard}`, "DELETE")
}

async function moveCard(idBoard, idCard, idListNow, idListDst, cix) {
    await fetchReq(`board/${idBoard}/card/${idCard}`, "PUT", {
        idListNow: idListNow,
        idListDst: idListDst,
        cix
    })
}

export default {
    createCard,
    getCard,
    updateCard,
    deleteCard,
    moveCard
}